tracker/enqueue.go

```diff
@@ -79,37 +95,90 @@ func (i *impl) TrackReference(ref Reference, obj interface{}) error {
 		"APIVersion": validation.IsQualifiedName(ref.APIVersion),
 		"Kind":       validation.IsCIdentifier(ref.Kind),
 		"Namespace":  validation.IsDNS1123Label(ref.Namespace),
-		"Name":       validation.IsDNS1123Subdomain(ref.Name),
 	}
+	var selector labels.Selector
 	fieldErrors := []string{}
+	switch {
+	case ref.Selector != nil && ref.Name != "":
+		fieldErrors = append(fieldErrors, "cannot provide both Name and Selector")
+	case ref.Name != "":
+		invalidFields["Name"] = validation.IsDNS1123Subdomain(ref.Name)
+	case ref.Selector != nil:
+		ls, err := metav1.LabelSelectorAsSelector(ref.Selector)
+		if err != nil {
+			invalidFields["Selector"] = []string{err.Error()}
+		}
+		selector = ls
+	default:
+		fieldErrors = append(fieldErrors, "must provide either Name or Selector")
+	}
 	for k, v := range invalidFields {
 		for _, msg := range v {
 			fieldErrors = append(fieldErrors, fmt.Sprintf("%s: %s", k, msg))
 		}
 	}
 	if len(fieldErrors) > 0 {
 		sort.Strings(fieldErrors)
-		return fmt.Errorf("invalid ObjectReference:\n%s", strings.Join(fieldErrors, "\n"))
+		return fmt.Errorf("invalid Reference:\n%s", strings.Join(fieldErrors, "\n"))
 	}
 
+	// Determine the key of the object tracking this reference.
 	object, err := kmeta.DeletionHandlingAccessor(obj)
 	if err != nil {
 		return err
 	}
-
 	key := types.NamespacedName{Namespace: object.GetNamespace(), Name: object.GetName()}
 
 	i.m.Lock()
 	defer i.m.Unlock()
-	if i.mapping == nil {
-		i.mapping = make(map[Reference]set)
+	if i.exact == nil {
+		i.exact = make(map[Reference]set)
+	}
+	if i.inexact == nil {
+		i.inexact = make(map[Reference]matchers)
+	}
+
+	// If the reference uses Name then it is an exact match.
+	if selector == nil {
+		l, ok := i.exact[ref]
+		if !ok {
+			l = set{}
+		}
+
+		if expiry, ok := l[key]; !ok || isExpired(expiry) {
+			// When covering an uncovered key, immediately call the
+			// registered callback to ensure that the following pattern
+			// doesn't create problems:
+			//    foo, err := lister.Get(key)
+			//    // Later...
+			//    err := tracker.Track(fooRef, parent)
+			// In this example, "Later" represents a window where "foo" may
+			// have changed or been created while the Track is not active.
+			// The simplest way of eliminating such a window is to call the
+			// callback to "catch up" immediately following new
+			// registrations.
+			i.cb(key)
```

Comment:
Although, looks like this was how it was before also, so not really your doing. Just curious if we really want to not unlock before making the callback.

Reply:
Still my fault, just from long ago :)

We could `go i.cb(key)`, but I think I prefer this since folks should really be doing fairly light work here (e.g. queue manipulation) vs. heavy listing (something that might make this re-entrant).

Reply:
Will follow up.